--- src/collectd.conf.in.orig	2013-01-27 14:47:18.246255910 +0100
+++ src/collectd.conf.in	2013-02-04 00:18:28.457298392 +0100
@@ -28,6 +28,7 @@
 
 #Timeout      2
 #ReadThreads  5
+Include      "/etc/collectd.d"
 
 ##############################################################################
 # Logging                                                                    #
@@ -136,7 +137,7 @@
 #@BUILD_PLUGIN_REDIS_TRUE@LoadPlugin redis
 #@BUILD_PLUGIN_ROUTEROS_TRUE@LoadPlugin routeros
 #@BUILD_PLUGIN_RRDCACHED_TRUE@LoadPlugin rrdcached
-@LOAD_PLUGIN_RRDTOOL@LoadPlugin rrdtool
+#@LOAD_PLUGIN_RRDTOOL@LoadPlugin rrdtool
 #@BUILD_PLUGIN_SENSORS_TRUE@LoadPlugin sensors
 #@BUILD_PLUGIN_SERIAL_TRUE@LoadPlugin serial
 #@BUILD_PLUGIN_SNMP_TRUE@LoadPlugin snmp
