--- bindings/Makefile.am.orig	2013-01-27 14:47:05.000000000 +0100
+++ bindings/Makefile.am	2013-02-04 00:48:01.854444442 +0100
@@ -45,7 +45,7 @@
 
 buildperl/Makefile: .perl-directory-stamp buildperl/Makefile.PL \
 	$(top_builddir)/config.status
-	cd buildperl && @PERL@ Makefile.PL INSTALL_BASE=$(DESTDIR)$(prefix) @PERL_BINDINGS_OPTIONS@
+	cd buildperl && @PERL@ Makefile.PL @PERL_BINDINGS_OPTIONS@
 
 buildperl/Makefile.PL: .perl-directory-stamp $(top_builddir)/config.status
 
--- bindings/Makefile.in.orig	2013-01-27 14:47:35.000000000 +0100
+++ bindings/Makefile.in	2013-02-04 00:48:51.440147674 +0100
@@ -716,7 +716,7 @@
 
 buildperl/Makefile: .perl-directory-stamp buildperl/Makefile.PL \
 	$(top_builddir)/config.status
-	cd buildperl && @PERL@ Makefile.PL INSTALL_BASE=$(DESTDIR)$(prefix) @PERL_BINDINGS_OPTIONS@
+	cd buildperl && @PERL@ Makefile.PL @PERL_BINDINGS_OPTIONS@
 
 buildperl/Makefile.PL: .perl-directory-stamp $(top_builddir)/config.status
 
